### ISSUE TITLE
request: earlier transfer of nightlogs

### ISSUE THREAD

**Issue by sbailey** [desihub/desitransfer#57]
**Problem**: In the current network situation, we don't get the emailed nightlogs, and the pdf posted to slack is just a printout of a webpage and truncates various scrollable tables.  It appears that the full HTML nightlogs are synced as part of the noontime daily engineering transfer to https://data.desi.lbl.gov/desi/survey/ops/nightlogs/, but this is later than we need them for the morning QA.

i.e. it used to be fine that those were synced later because we had the email version earlier, but now it would be quite useful to have them sooner.

**Request**: could we sync these earlier, e.g. at 7:30 MST (~sunrise at the winter solstice)?

As an implementation detail, perhaps this could be implemented by splitting the daily transfer into two pieces: the standard noontime script on one set of directories, and a different copy that does the same thing but on a different set of directories earlier in the day.  Candidates for earlier transfers:
  * nightlogs
  * spacewatch (#54)
  * gfaproc (until we finish porting those from KPNO to NERSC)

**Comment by weaverba137** [comment]
@sbailey, this is an interesting idea. I am pretty sure we could split the daily transfers this way. Same code, different configuration, with the configuration selected by a command-line argument.

**Comment by weaverba137** [comment]
Although there is a catch: with spacewatch, we would need to use http, ultimately via `requests`. We can't use rsync there, and all the other daily transfers are built around rsync.

**Comment by weaverba137** [comment]
And since GFA processing will be moved to NERSC, then the entirety of any alternate time implementation would only ever be for the benefit nightlog data.

Is it really, truly impossible to send email from the DESI cluster, or has it just not been set up yet?

**Comment by daqiii** [comment]
Stephen, All
What exactly do you need that’s not available in the nightlog? Is it the list of exposures? That you can easily retrieve from the NERSC DB mirror For example, for last night:

***@***.***

And if you need more detail information on a particular exposure, that is available too. For example


***@***.***

Cheers,
Klaus

From: Stephen Bailey ***@***.***>
Date: Monday, October 30, 2023 at 5:28 PM
To: desihub/desitransfer ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [desihub/desitransfer] request: earlier transfer of nightlogs (Issue #57)
Problem: In the current network situation, we don't get the emailed nightlogs, and the pdf posted to slack is just a printout of a webpage and truncates various scrollable tables. It appears that the full HTML nightlogs are synced as part of


Problem: In the current network situation, we don't get the emailed nightlogs, and the pdf posted to slack is just a printout of a webpage and truncates various scrollable tables. It appears that the full HTML nightlogs are synced as part of the noontime daily engineering transfer to https://data.desi.lbl.gov/desi/survey/ops/nightlogs/<https://urldefense.com/v3/__https:/data.desi.lbl.gov/desi/survey/ops/nightlogs/__;!!KGKeukY!20jQFtuF4CTSRvFFWhj63ap-TS9Wu4-QLk9gDzmJtRSXTrjqXkX7WQ9TjQUakSYFBJKF6Oy-SZI042_hQYUYaVjJlg$>, but this is later than we need them for the morning QA.

i.e. it used to be fine that those were synced later because we had the email version earlier, but now it would be quite useful to have them sooner.

Request: could we sync these earlier, e.g. at 7:30 MST (~sunrise at the winter solstice)?

As an implementation detail, perhaps this could be implemented by splitting the daily transfer into two pieces: the standard noontime script on one set of directories, and a different copy that does the same thing but on a different set of directories earlier in the day. Candidates for earlier transfers:

  *   nightlogs
  *   spacewatch
  *   gfaproc (until we finish porting those from KPNO to NERSC)

**Comment by weaverba137** [comment]
@daqiii, unfortunately your comment was severely garbled. I think you will have to log in to GitHub and comment that way.

**Comment by weaverba137** [comment]
But I completely agree that I should not be duplicating effort if all the information is already available via the replicator or telemetry interfaces.

**Comment by daqiii** [comment]
Thanks Ben – I think I made my point. What’s missing are two screen shots from DB that anyone can easily reproduce. Here is the URL of the exposure table viewer at  NERSC:

https://replicator.desi.lbl.gov/EXPO/Expo.html

Klaus

From: Benjamin Alan Weaver ***@***.***>
Date: Monday, October 30, 2023 at 5:56 PM
To: desihub/desitransfer ***@***.***>
Cc: Honscheid, Klaus ***@***.***>, Mention ***@***.***>
Subject: Re: [desihub/desitransfer] request: earlier transfer of nightlogs (Issue #57)
But I completely agree that I should not be duplicating effort if all the information is already available via the replicator or telemetry interfaces. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving


But I completely agree that I should not be duplicating effort if all the information is already available via the replicator or telemetry interfaces.

**Comment by daqiii** [comment]
We will be able to send email from the DESI cluster – we also need that for critical alerts. But this will only be enabled when we are back on the regular network. (unless someone makes a very strong case that this is absolutely needed now and that we need to find a way to achieve this via starlink)

Klaus

From: Benjamin Alan Weaver ***@***.***>
Date: Monday, October 30, 2023 at 5:38 PM
To: desihub/desitransfer ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [desihub/desitransfer] request: earlier transfer of nightlogs (Issue #57)
And since GFA processing will be moved to NERSC, then the entirety of any alternate time implementation would only ever be for the benefit nightlog data. Is it really, truly impossible to send email from the DESI cluster, or has it just not


And since GFA processing will be moved to NERSC, then the entirety of any alternate time implementation would only ever be for the benefit nightlog data.

Is it really, truly impossible to send email from the DESI cluster, or has it just not been set up yet?

**Comment by sbailey** [comment]
@akremin and @abhi0395 have been running data pipeline daily ops recently so they are much more familiar with the kinds of things that they miss from the old email logs; I'm mainly just an incomplete messenger here.  IIUC, sometimes the pdf printout appears truncated (which does seem to be the case for the "all exposures" table at the end; I'm not sure if there are other cases within the log).  Sometimes the info is there but partially obscured, adding another layer of friction when debugging something:
<img width="603" alt="image" src="https://github.com/desihub/desitransfer/assets/218471/a23822ef-55c6-444c-9a81-13923bffa9cc">.  Other times it might be that the log wasn't posted to slack in time.

Although the desired information may be available by other means, using other interfaces or code to access it, the idea behind this ticket is that the file we want is already on disk at KPNO and we already have the infrastructure to grab it at noon, so could we just grab it at 7:30 instead?

That being said, I agree that gfaproc is hopefully only a temporary need (though perhaps we should have implemented this "other times" flexibility years ago), and good point that spacewatch is already a special case different from rsync.  But even if the nightlog is the only case, adding a sync for that at 7:30 still seems like it might be relatively simple to do, and simpler than developing other solutions.

**Comment by akremin** [comment]
One other reason I might add is that it is very helpful to have a
searchable medium to quickly look for all log entries related to exposure
XXXX or tile YYYY without needing to open individual PDF's (in email this
was from the search bar and in html files it is grep-able).

The other reasons are exactly as Stephen mentioned: 1) Being able to find
both the LO/SO comments for exposures and a complete list of exposures in
one location is convenient without having to query a database. 2) There are
also entries that get cut off as shown in Stephen's screenshot.

None of these issues are insurmountable, but the current situation does add
time and a level of inconvenience to the daily data ops tasks.

If the fix is difficult, then we don't need to do anything until the full
network configuration is ready. But if it is easy and straightforward
to add an extra transfer, then it will be a net gain in productivity.

**Comment by weaverba137** [comment]
Unfortunately, things have become more confusing to me rather than clearer. Please note that the comments below reflect my genuine unfamiliarity with daily pipeline operations.

1. We've been operating in a mode where nightlog emails have been unavailable since August. Why wasn't this request made the moment that line internet was restored in late September? Or even earlier, so that we could implement it as soon as transfers were restored?
2. @daqiii, do you know approximately when outgoing email capability will be restored? If it's on the order of a week or two, that's the same timescale on which any implementation of this request could be deployed. The implementation timescale will probably be pushed out even further by higher-priority tasks related to DR1.
3. We have a database replica at NERSC, and if that isn't a "searchable medium", I don't know what is. Presumably that should be the authoritative source for what exposures were taken. Is the pipeline team really using the replicator database system to the maximum extent? If not, why not? This is a serious concern because maintenance of that database is non-trivial. If the database isn't used, those resources should be freed up for other tasks.
4. There is an additional maintenance headache: the cron jobs that perform the existing transfers operate on Pacific Time, so the time at which the existing daily transfer runs has to be adjusted twice a year to keep it in sync with noon MST.  By adding an additional transfer at 7:30 MST, that would be two separate cron jobs that would need to be adjusted twice a year.
5. Finally, as I mentioned, my knowledge has gaps. What is the document you suggest people read when onboarding them for daily pipeline operations?

**Comment by sbailey** [comment]
Unfortunately, things have become more confusing to me rather than clearer. Please note that the comments below reflect my genuine unfamiliarity with daily pipeline operations.

> 1. We've been operating in a mode where nightlog emails have been unavailable since August. Why wasn't this request made the moment that line internet was restored in late September? Or even earlier, so that we could implement it as soon as transfers were restored?

I wish we had requested this much earlier.  Last week I reviewed with Anthony and Abhijeet why dailyops was taking so much effort, and getting the full html nightlog earlier in the day was identified as something that would help, thus this ticket.  And in general this flexibility is something that we also would have used for gfaproc if it had been available before, and I expect cases like this will come up again in the future.

> 2. @daqiii, do you know approximately when outgoing email capability will be restored? If it's on the order of a week or two, that's the same timescale on which any implementation of this request could be deployed. The implementation timescale will probably be pushed out even further by higher-priority tasks related to DR1.

(I'm not @daqiii, but answering from my understanding): From the instrument call on Monday, it sounds like that won't happen for at least another month ("back on the regular network" won't happen until Nov bright time), and IIUC the email server reconfig may take additional work beyond that.

> 3. We have a database replica at NERSC, and if that isn't a "searchable medium", I don't know what is. Presumably that should be the authoritative source for what exposures were taken. Is the pipeline team really using the replicator database system to the maximum extent? If not, why not? This is a serious concern because maintenance of that database is non-trivial. If the database isn't used, those resources should be freed up for other tasks.

I think this is getting off topic for this ticket.  The replicator is used for the graphana display and I think by individual analyses such as focal plane work.  What we are asking for here is the night log, with all of its information and commentary, which is a different use case.  Being able to grep for all mentioned of a given tileid, or "VCCD" or "stray light" can be useful in a way that is different from the usefulness of an exposure database.

> 4. There is an additional maintenance headache: the cron jobs that perform the existing transfers operate on Pacific Time, so the time at which the existing daily transfer runs has to be adjusted twice a year to keep it in sync with noon MST. By adding an additional transfer at 7:30 MST, that would be two separate cron jobs that would need to be adjusted twice a year.

It doesn't have to be exactly 7:30 MST.  8am PT would be fine too.  Just something earlier in the day, preferably well before 9am.

> 5. Finally, as I mentioned, my knowledge has gaps. What is the document you suggest people read when onboarding them for daily pipeline operations?

https://desi.lbl.gov/trac/wiki/Pipeline/DailyOps

Checking the nightlog is the first of the "Morning Steps to Sign off on a night", and more implicitly it is used to understand what happened when evaluating anything that went wrong with the data processing (e.g. if PSF fits are failing, checking the logs to see if there is any commentary about lights being on, oddities seen by the support observer in data quality, etc).

**Comment by akremin** [comment]
Just elaborating a little bit more on (1), (3), and (5), which will
hopefully add to the clarification.

1) We have a pdf printout of the nightlog sent to us by the LO's each
morning at the end of the night, but it is far from ideal for the reasons
mentioned earlier. We have lived with it assuming it was a temporary
inconvenience, but it is now clear that we won't be back to emails for a
while yet. The inconvenience has also grown into more of a problem since
actual mistakes have been made on data management tasks, which elevated the
desire to look for a better solution.

3) We want to be able to search the SO and LO log messages for each
exposure/tile to get information from them about what was happening at the
telescope at the time the data was taken, and their data quality
assessments of each exposure based on the QA plots.

5) As Stephen said, you can pick a single time. Maybe the latest sunrise of
the year in Tucson?

**Comment by weaverba137** [comment]
Thank you @sbailey and @kremin. I will proceed with getting this set up.

**Comment by abhi0395** [comment]
Thanks @weaverba137 @sbailey @akremin, Yes I agree with @akremin. Recent PDFs of nightlogs have been difficult to read and go through exposures. So it will be great to have the proper nightlogs we used to get in the past. For example, the night log for 1028 is missing many exposures and their details, and it was particularly a difficult night for processing. .

**Comment by weaverba137** [comment]
@abhi0395, That's interesting, because the nightlog data for 20231028 should have been transferred already. If the observers did not record details of the exposures, then transferring nightlog data early would not help. Could you elaborate on what happened?

**Comment by abhi0395** [comment]
@weaverba137, the night was particularly problematic; first, there were some spectrograph issues, so the evening cals/darks had to be discarded. However, there are quite a few dark exposures that were possibly taken between the science tiles (I am not sure, just guessing from expids), but those exposures are missing in the nightlog. Then, a new set of zeros, flats and arcs was taken in the morning, but those exposures are also missing the log, and there is not much discussion about them. A person like me who tries to cross-verify things between nightlogs and nightwatch (at NERSC, I don't have much idea about KPNO nightwatch), it was difficult to verify things as nightwatch was also not updated for that night. 

To be more specific, I had to go through the raw data folder and exposure tables to understand the expid of a new set of cals. For example, in the nightlog, there are no exposures between 202424 and 202446. Also, there is no mention of exposures after 202457. Also since the pdf is truncated it's not a helpful table.

**Comment by weaverba137** [comment]
@abhi0395, I'm still confused. Unless I'm misunderstanding something, you're basically saying that having the nightlog data earlier would not help at all because observers are not writing good logs. Could you please clarify?

**Comment by abhi0395** [comment]
@weaverba137 I am also unaware of how the nightlogs are written and generated. If what I am saying is mostly human things, then having the nightlog earlier won't be useful unless it's well-written. I also want to understand if the final exposure table that's part of the nightlog is generated automatically; if yes, then having a complete exposure table in the nightlog would be very useful in case there is a missing exposure in the exposure-wise detail.

**Comment by akremin** [comment]
I'm not sure why it didn't populate properly yesterday or where it
populates from.

But the key point is that the list of exposures is a distant third in the
list of motivations for this request. We could in theory get the exposure
list elsewhere. What we can't get are the other things. The list of
exposures in the same location is just a nice additional benefit that is
very useful when it is available (which it is most of the time).

tl;dr We still want the nightlogs even if the issue that happened last
night happens occasionally.

**Comment by abhi0395** [comment]
@akremin yes, I agree that having the proper nightlogs we used to get in the past would be very useful. I also meant the same :)
